docs/design.md

```diff
@@ -1,175 +1,241 @@
-3. Experiment with the model and training and only save my notebook somewhere for reference
+In this case, the `notebook-gpu` image would have all the `transformers`, `pytorch`, `cuda`, `python 3`, etc. libraries pre-installed.
 
 ```yaml
 kind: Notebook
-name: notebooks-for-anything
+metadata:
+  name: notebooks-for-anything
 spec:
-  container:
-    # or git repo as a source
-    image: substratusai/notebook-gpu
-  resources: {gpuCount: 4, type: L4}
+  image:
+    name: substratusai/notebook-gpu
+  resources:
+    gpu:
+      count: 4
+      type: nvidia-l4
 ```
 
+#### Usecase: Iterate on Model Training Code
+
+In this case, a user might want to update the code used for training in a Model object. The goal here is to create a development environment for the user that exactly mimics the training environment.
 
-Lab user flow: Iterate on model training code using a customized trainer image
+Steps:
 
 1. `git clone https://substratusai/hf-llm-trainer && cd hf-llm-trainer`
-2. Modify Dockerfile
-3. Create a Notebook manifest and apply the Notebook CR mounting a source model
-4. `kubectl open notebook -f notebook-training-experiment` - the `kubectl` plugin does each of the following additional steps when `spec.container.local = true`
-   * tar up directory respecting dockerignore
-   * Get GCS signed url from controller notebook status 
-   * upload to workspace.tar to signed URL
-5. Notebook automatically has the local directory contents in the notebook when using the `kubectl` plugin (via cp sync operations)
+2. [Optionally] Modify Dockerfile.
+3. `kubectl open notebook -f .`
+4. The kubectl plugin does the following:
+   * Creates a Notebook with `.image.upload` set.
+   * Tars local directory respecting `.dockerignore`
+5. A Substratus controller in the background:
+   * Creates a signed URL for the upload.
+   * Updates the Notebook status with the signed URL.
+6. The kubectl plugin continues:
+   * Uploads tar to signed URL.
+7. The Substratus controller is now orchestrating the build of this image using kaniko.
+8. After the Notebook is marked as Ready (`.status.ready: true`), the kubectl plugin:
+   * Copies local `*.py` files into the running notebook Pod.
+   * Fetches a token reported in the Notebook `.status.token` field.
+   * Opens browser to `http://localhost:8888?with-token=...`.
 
 ```yaml
 kind: Notebook
 name: notebook-training-experiment
 spec:
-  container:
-    local: true
-  sourceModel: falcon-7b
-  resources: {}
+  image:
+    upload: {} # This is how the plugin signals it wants to upload a directory for building.
+  model: falcon-7b # Mounts the model. Plugin autopopulated this by finding the corresponding `model.yaml` file.
+  resources: {...} # Plugin autopopulated this by finding the corresponding `model.yaml` file.
 status:
-    uploadUrl: gs://substratusai-notebooks/workspace.tar
+  uploadUrl: https://some-signed-url... # Controller populated this.
+  token: aklsdjfkasdljfs # Reported by the controller.
```

Comment:
Remind me what this is again?

Reply:
Notebook token that it complains about when you don’t use one